### ISSUE TITLE
Question about Kuka Training and Kuka Dataset.

### ISSUE THREAD

**Issue by erdiphd** [jannerm/diffuser#32]
Hi.

I am using the _kuka_ branch and downloaded the dataset from this link https://www.dropbox.com/s/zofqvtkwpmp4v44/metainfo.tar.gz?dl=0

I added the following line to the script to see the visualization of the datasets: ( more specifically, I added them in the line 71 https://github.com/jannerm/diffuser/blob/kuka/scripts/kuka.py#L71)

```
tmp = np.load("kuka_dataset/data_897.npy")
renderer.renders(tmp)
```
And the visualization is the following video

https://user-images.githubusercontent.com/81103451/226195002-4ab2ef7b-1224-4e48-b0ca-c5eef5bb1362.mp4

As seen in the video, the robot has already picked up and placed the object. 

**1) My question is:**

What is the objective of training after loading the dataset?

Then, I continued training until the timestep is **15000** ( the loss is around 0.16). 

I believe you sample the images(trajectories) starting from the this line https://github.com/jannerm/diffuser/blob/kuka/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L800
and replay the robot.
![image](https://user-images.githubusercontent.com/81103451/226196013-90dc66d5-f885-4681-a3c7-44222ed909b9.png)

I wanted to see what would look like the robot movement after training the neural network. Then I sampled the images from this line and replay the robot. The robot as well as the cube movements can be seen in the following video?

https://user-images.githubusercontent.com/81103451/226196095-0795d5c1-0470-4e0a-980c-c357f9150713.mp4

**2) My second question** is why the behaviour is like that ?

**Comment by yilundu** [comment]
Hi sorry about the late reply -- the goal of the model is to learn pick and place one single block (given other already stacked blocks)

In the second setting -- the model predicts when to form contacts with individual blocks. When the model is not trained well, the contacts will be incorrectly added, which leads to very odd physics simulation.

Feel free to let me know if you have any other questions.

**Comment by erdiphd** [comment]
@yilundu , Thank you very much for you reply. Unfortunately, I didn't understand.

Let me write my question in little bit detail. 

### The part before training

You have this kind of dataset , _kuka_dataset/data_897.npy_, _kuka_dataset/data_898.npy_ .... and diffuser gets this dataset and trains the model.

When I check your dataset (e.g kuka_dataset/data_897.npy) , the dataset consists of 470x39 parameters.

(The first axis of the dataset) The 470 is timestep **t**  
(The second axis of the dataset) 7 variables from the 39 dimension are the joint variables of the robot. The rest of the parameters (32 parameters) is the position as well as the orientation of the 4 cubes ( 3 pos and 4 rot for each cube)

Basically, the data has the trajectory points of each cube as well as the joint values for the robot in each timestep.

If I render the following datasets (without starting any training), I can already see that the robot picks objects and places them. ( In the following video)

```
tmp = np.load("kuka_dataset/data_897.npy")
renderer.renders(tmp)
```
Here is the rendering of the dataset **data_897**

https://user-images.githubusercontent.com/81103451/227969884-bcbeebc7-f814-425f-a444-cce1a648a31e.mp4

```
tmp = np.load("kuka_dataset/data_898.npy")
renderer.renders(tmp)
```
Here is the rendering of the dataset **data_898**
https://user-images.githubusercontent.com/81103451/227969534-5ec34a13-8f73-4c9a-ae01-4c5e204db0ea.mp4

So, the data was literally trained, and the robot learned how to pick and place. You start with a learned policy.

### The part after training
Then I start training using the pretrained model _model-650.pt_, then the error is around **0.020**
Then diffision model samples the trajectories points from the below code line ( here https://github.com/jannerm/diffuser/blob/kuka/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L806 )
![image](https://user-images.githubusercontent.com/81103451/227971513-cd3fa1ee-34cc-425b-b96e-fd6eca46369a.png)

When I check the dimension of the variable _unnormed_, It has size 
```
unnormed.shape
torch.Size([1, 128, 39])
```
It means that It has joint values as well as pos and rot for each cube for 128 timesteps. When I render this data ( the sampled data from the diffusion model), I see the following behaviour;



https://user-images.githubusercontent.com/81103451/227972662-60d2c490-8707-4950-a6b0-5a870ea1fda9.mp4

So, It is simply the similar behaviour before starting to the training.


You are giving some data from a learned policy(e.g robot pick and place the cube). The diffuser model gets trajectory points of the cube as well as joints of the robot, It trains it and obtains some trajectory points(still robot picks and place the cube). ( If I understand correctly)

So my question is;

The robot  already picks and place the object before training. After the training, It has still the same behavior. What is the objective of the training?

**Comment by yilundu** [comment]
Ahh I see -- yes the objective of training is to learn how to pick / place blocks from a large dataset of existing demonstrations. The goal is to generalize to new configuration of blocks seen at test time, as well as pick and place blocks in a controllable manner.
